File: carmajava/mock_drivers/src/main/java/gov/dot/fhwa/saxton/carma/mock_drivers/MockLightBarDriver.java
/*
 * Copyright (C) 2018-2020 LEIDOS.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package gov.dot.fhwa.saxton.carma.mock_drivers;

import org.ros.message.Time;
import org.ros.node.ConnectedNode;
import org.ros.node.topic.Publisher;
import org.ros.exception.ServiceException;
import org.ros.node.service.ServiceResponseBuilder;
import org.ros.node.service.ServiceServer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import cav_msgs.LightBarStatus;
import cav_srvs.GetLightsRequest;
import cav_srvs.GetLightsResponse;
import cav_srvs.SetLightsRequest;
import cav_srvs.SetLightsResponse;


/**
 * A class which can be used to simulate a LightBar driver for the CARMA Platform.
 * <p>
 * Command line test:
 * ROSJava does not support rosrun parameter setting so a rosrun is a multi step process
 * rosparam set /mock_driver/simulated_driver 'pinpoint'
 * rosparam set /mock_driver/data_file_path '/opt/carma/test_data/pinpoint_stationary.csv'
 * rosrun carma mock_drivers gov.dot.fhwa.saxton.carma.mock_drivers.MockDriverNode
 */
public class MockLightBarDriver extends AbstractMockDriver {

  // Topics
  // Published
  final Publisher<cav_msgs.LightBarStatus> lbPub;
  final String lbStatusTopic = "lightbar/light_bar_status";

  // Services
  // Server
  protected final ServiceServer<cav_srvs.SetLightsRequest, cav_srvs.SetLightsResponse> setLightsService;
  protected final ServiceServer<cav_srvs.GetLightsRequest, cav_srvs.GetLightsResponse> getLightsService;
  final String lbGetLightsService = "lightbar/get_lights";
  final String lbSetLightsService = "lightbar/set_lights";  

  private final short EXPECTED_DATA_COL_COUNT = 132; // TODO
  private final short SAMPLE_ID_IDX = 1;

  // LightBar States
  protected boolean greenFlash = false;
  protected boolean yellowFlash = false;
  protected boolean leftArrow = false;
  protected boolean rightArrow = false;
  protected boolean sidesSolid = false;
  protected boolean greenSolid = false;
  protected boolean yellowDim = false;
  protected boolean takedown = false;


  public MockLightBarDriver (ConnectedNode connectedNode) {
    super(connectedNode);
    // Topics
    // Published
    lbPub =
      connectedNode.newPublisher(lbStatusTopic, cav_msgs.LightBarStatus._TYPE);

    // Services
    // Server
    setLightsService = connectedNode.newServiceServer(lbSetLightsService, cav_srvs.SetLights._TYPE,
        new ServiceResponseBuilder<SetLightsRequest, SetLightsResponse>() {
          @Override
          public void build(cav_srvs.SetLightsRequest request, cav_srvs.SetLightsResponse response) {

            cav_msgs.LightBarStatus lightStatus = request.getSetState();
            greenFlash = lightStatus.getGreenFlash() == 1;
            yellowFlash = lightStatus.getFlash() == 1;
            leftArrow = lightStatus.getLeftArrow() == 1;
            rightArrow = lightStatus.getRightArrow() == 1;
            sidesSolid = lightStatus.getSidesSolid() == 1;
            greenSolid = lightStatus.getGreenSolid() == 1;
            yellowDim = lightStatus.getYellowSolid() == 1;
            takedown = lightStatus.getTakedown() == 1;
  
            log.info("Lights have been set to: " + 
            "\nGreen Flash: " + greenFlash + 
            "\nYellow Flash: " + yellowFlash + 
            "\nLeft: " + leftArrow + 
            "\nRight: " + rightArrow + 
            "\nSides Solid: " + sidesSolid + 
            "\nGreen Solid: " + greenSolid + 
            "\nYellow Dim: " + yellowDim + 
            "\nTakedown: " + takedown);
          }
        });

    getLightsService = connectedNode.newServiceServer(lbGetLightsService, cav_srvs.GetLights._TYPE,
        new ServiceResponseBuilder<GetLightsRequest, GetLightsResponse>() {
          @Override
          public void build(cav_srvs.GetLightsRequest request, cav_srvs.GetLightsResponse response) {
          
          response.setStatus(getLightBarStatus());

          }
        });
  }

  /**
   * Helper function to build the lightbar status message
   * @return The lightbar status message
   */
  protected LightBarStatus getLightBarStatus() {
    cav_msgs.LightBarStatus lightStatus = lbPub.newMessage();

    lightStatus.setGreenSolid((byte)(greenSolid?1:0));
    lightStatus.setYellowSolid((byte)(yellowDim?1:0));
    lightStatus.setRightArrow((byte)(rightArrow?1:0));
    lightStatus.setLeftArrow((byte)(leftArrow?1:0));
    lightStatus.setSidesSolid((byte)(sidesSolid?1:0));
    lightStatus.setFlash((byte)(yellowFlash?1:0));
    lightStatus.setGreenFlash((byte)(greenFlash?1:0));
    lightStatus.setTakedown((byte)(takedown?1:0));

    return lightStatus;
  }
  
  @Override protected void publishData(List<String[]> data) throws IllegalArgumentException {
    // TODO use actual data from file
    cav_msgs.LightBarStatus lightStatus = getLightBarStatus();
  
    lbPub.publish(lightStatus);
  }

  @Override protected short getExpectedColCount() {
    return EXPECTED_DATA_COL_COUNT;
  }

  @Override protected short getSampleIdIdx(){
    return SAMPLE_ID_IDX;
  }

  @Override protected List<String> getDriverTypesList() {
    return new ArrayList<>(Arrays.asList("lightbar"));
  }

  @Override public List<String> getDriverAPI() {
    return new ArrayList<>(Arrays.asList(
      lbStatusTopic,
      lbGetLightsService,
      lbSetLightsService));
  }
}

File: translated_mock_drivers/translated_mock_drivers/MockLightBarDriver.cpp
#include <ros/ros.h>
#include <ros/node_handle.h>
#include <iostream>
#include <ros/service_client.h>
#include <ros/service_server.h>
#include "IMockDriver.cpp"
#include "AbstractMockDriver.cpp"
#include <cav_srvs/GetLights.h>
#include <cav_srvs/GetLightsRequest.h>
#include <cav_srvs/GetLightsResponse.h>
#include <cav_srvs/SetLights.h>
#include <cav_srvs/SetLightsRequest.h>
#include <cav_srvs/SetLightsResponse.h>
#include <cav_msgs/LightBarStatus.h>
using namespace std;

namespace mock_drivers{
    class MockLightBarDriver : public AbstractMockDriver {
         //topics published
         ros::Publisher lbPub;
         string lbStatusTopic = "lightbar/light_bar_status";

         //service server
         ros::ServiceServer getLightsService;
         ros::ServiceServer setLightsService;
         string lbGetLightService = "lightbar/get_lights";
         string lbSetLightService = "lightbar/set_lights";
         short EXPECTED_DATA_COL_COUNT = 132;
         short SAMPLE_ID_IDX = 1;

         //lightbar states
          bool greenFlash = false;
          bool yellowFlash = false;
          bool leftArrow = false;
          bool rightArrow = false;
          bool sidesSolid = false;
          bool greenSolid = false;
          bool yellowDim = false;
          bool takedown = false;


          public:
            MockLightBarDriver(ros::NodeHandle node) : AbstractMockDriver(node) {
                //topics published
                lbPub = node.advertise<cav_msgs::LightBarStatus>(lbStatusTopic, 100);

                //service server
                setLightsService = node.advertiseService(lbSetLightService, &MockLightBarDriver::setLightService_cb, this);
                getLightsService = node.advertiseService(lbGetLightService, &MockLightBarDriver::getLightService_cb, this);
            }

            //set light service call back
            bool setLightService_cb(cav_srvs::SetLightsRequest& req, cav_srvs::SetLightsResponse& resp){
                cav_msgs::LightBarStatus lightstatus = req.set_state;
                greenFlash = lightstatus.green_flash == 1;
                yellowFlash = lightstatus.flash == 1;
                leftArrow = lightstatus.left_arrow == 1;
                rightArrow = lightstatus.right_arrow== 1;
                sidesSolid = lightstatus.sides_solid == 1;
                greenSolid = lightstatus.green_solid == 1;
                yellowDim = lightstatus.yellow_solid == 1;
                takedown = lightstatus.takedown == 1;
                return true;
            }

            //get light service call back
            bool getLightService_cb(cav_srvs::GetLightsRequest& req, cav_srvs::GetLightsResponse& resp){
                resp.status = getLightBarStatus();
                return true;
            }

            /**
            * Helper function to build the lightbar status message
            * @return The lightbar status message
            */
            cav_msgs::LightBarStatus getLightBarStatus(){
                cav_msgs::LightBarStatus status;
                status.green_solid = greenSolid? 1:0;
                status.yellow_solid = yellowDim? 1:0;
                status.right_arrow = rightArrow? 1:0;
                status.left_arrow = leftArrow? 1:0;
                status.sides_solid = sidesSolid? 1:0;
                status.flash = yellowFlash? 1:0;
                status.green_flash = greenFlash? 1:0;
                status.takedown = takedown? 1:0;
                return status;

            }

            vector<string> getDriverApi() override {
                vector<string> api;
                api.push_back(lbStatusTopic);
                api.push_back(lbGetLightService);
                api.push_back(lbSetLightService);
                return api;

            } 
        
            void publishData(list<vector<string>> data) override {
                cav_msgs::LightBarStatus status = getLightBarStatus();
                lbPub.publish(status);
            }

            short getExpectedColCount() override {
                return EXPECTED_DATA_COL_COUNT;
            }

            short getSampleIdIdx() override {
                return SAMPLE_ID_IDX;
            }

            vector<string> getDriverTypesList() override {
                vector<string> list;
                list.push_back("lightbar");
                return list;
            }
    };

}